src/bidiMapper/domains/context/context.ts

```diff
@@ -219,187 +136,127 @@ export class Context {
     _this: Script.ArgumentValue,
     args: Script.ArgumentValue[],
     awaitPromise: boolean
   ): Promise<Script.CallFunctionResult> {
-    const cdpCallFunctionResult = await this._executeCallFunction(
+    return this.#scriptEvaluator.callFunction(
       functionDeclaration,
       _this,
       args,
       awaitPromise
     );
-    if (cdpCallFunctionResult.exceptionDetails) {
-      // Serialize exception details.
-      return await this._serializeCdpExceptionDetails(
-        cdpCallFunctionResult.exceptionDetails,
-        this._callFunctionStacktraceLineOffset
-      );
-    }
-
-    return {
-      result: cdpCallFunctionResult.result.value,
-    };
   }
 
   public async findElement(
     selector: string
-  ): Promise<CommonDataTypes.NodeRemoteValue> {
-    const functionDeclaration =
-      '(resultsSelector) => document.querySelector(resultsSelector)';
-    const args: ArgumentValue[] = [{ type: 'string', value: selector }];
+  ): Promise<BrowsingContext.PROTO.FindElementResult> {
+    const functionDeclaration = String((resultsSelector: string) =>
+      document.querySelector(resultsSelector)
+    );
+    const args: Script.ArgumentValue[] = [{ type: 'string', value: selector }];
 
-    const findElementCommandResult = await this._executeCallFunction(
+    return (await this.#scriptEvaluator.callFunction(
       functionDeclaration,
       {
         type: 'undefined',
       },
       args,
       true
-    );
+    )) as BrowsingContext.PROTO.FindElementResult;
+  }
 
-    if (findElementCommandResult.exceptionDetails) {
-      // Serialize exception details.
-      throw await this._serializeCdpExceptionDetails(
-        findElementCommandResult.exceptionDetails,
-        this._callFunctionStacktraceLineOffset
+  async #initialize() {
+    // Enabling Runtime doamin needed to have an exception stacktrace in
+    // `evaluateScript`.
+    await this.#cdpClient.Runtime.enable();
+    await this.#cdpClient.Page.enable();
+    await this.#cdpClient.Page.setLifecycleEventsEnabled({ enabled: true });
+    this.#initializeEventListeners();
+  }
+
+  #initializeEventListeners() {
+    this.#initializePageLifecycleEventListener();
+    this.#initializeCdpEventListeners();
+    // TODO(sadym): implement.
+    // this.#handleBindingCalledEvent();
+  }
+
+  #initializeCdpEventListeners() {
+    this.#cdpClient.on('event', async (method, params) => {
+      await this.#eventManager.sendEvent(
+        {
+          method: 'PROTO.cdp.eventReceived',
+          params: {
+            cdpMethod: method,
+            cdpParams: params,
+            session: this._sessionId,
+          },
+        },
+        null
       );
-    }
+    });
+  }
 
-    return findElementCommandResult.result
-      .value as CommonDataTypes.NodeRemoteValue;
+  #initializePageLifecycleEventListener() {
+    this.#cdpClient.Page.setLifecycleEventsEnabled({ enabled: true });
+
+    this.#cdpClient.Page.on('lifecycleEvent', async (params) => {
```

Comment:
should there be .off calls for every .on call somewhere?

Reply:
Let's keep it out of scope of this PR. Ideally, we should subscribe to events only when user subscribed to them, and unsubscribe when user unsubscribed. But for now Mapper (as well as Puppeteer, AFAIK) just listens to all event and sends the needed to user.